{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# How do I _copy_ an app from the Seven Bridges _Public Apps_ repository?\n",
    "## Overview\n",
    "Apps include both tools (individual bioinformatics utilities) and workflows (chains or pipelines of connected tools). Apps are located in one of two places: the [Seven Bridges Public Apps repository](https://igor.sbgenomics.com/public/apps) or a user-created project.\n",
    "\n",
    "The Seven Bridges Platform offers a repository of [publicly available apps](https://igor.sbgenomics.com/public/apps) suitable for many different types of data analysis. Apps include both tools (individual bioinformatics utilities) and workflows (chains or pipelines of connected tools). The publicly available apps are maintained by the Seven Bridges Platform bioinformatics team to represent the latest tool versions.\n",
    "\n",
    "## Objective\n",
    "\n",
    "You can copy an app from the Public App repository or from a project of which you're a member. Here, we focus on **copying an app from the Public App repository**. [[reference](http://docs.sevenbridges.com/docs/copy-an-app)]\n",
    " \n",
    "\n",
    "## Procedure\n",
    "1. We'll list all our projects.\n",
    "2. Then, we'll list the apps within the Seven Bridges Public Apps repository. \n",
    "3. Lastly, we'll copy an app from the Public Apps repository to the first project in our list of projects.\n",
    "\n",
    "## Prerequisites\n",
    " 1. You need to be a member (or owner) of at least one project. Learn more about creating a new project. [[recipe](projects_makeNew.ipynb)] [[reference](http://docs.sevenbridges.com/docs/create-a-new-project)].\n",
    " 2. You need your **authentication token** and the API needs to know about it. See <a href=\"Setup_API_environment.ipynb\">Setup_API_environment.ipynb</a> for details. Learn more about [obtaining your authentication token](http://docs.sevenbridges.com/v1.0/docs/get-your-authentication-token).\n",
    " 3. You understand how to list projects of which you are a member. We will use this call directly to select a project below. [[recipe](projects_listAll.ipynb)] [[reference](http://docs.sevenbridges.com/docs/list-all-your-projects)]\n",
    " 4. You understand how to list apps within a project. We will use this call below and select an app. [[recipe](apps_listAll.ipynb)] [[reference](http://docs.sevenbridges.com/docs/list-all-apps-available-to-you)]\n",
    " \n",
    "Note that as with all okAPI recipes, this recipe makes use of the [Seven Bridges Public API Python library](../Tutorials/SBPLAT/Setup_API_environment.ipynb)\n",
    " \n",
    "## Imports\n",
    "We import the `Api` class from the official `sevenbridges-python` bindings below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import sevenbridges as sbg"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initialize the object\n",
    "The _Api_ object needs to know your **auth\\_token** and the correct path. Here we assume you are using the .sbgrc file in your home directory. For other options see <a href=\"Setup_API_environment.ipynb\">Setup_API_environment.ipynb</a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# [USER INPUT] specify platform {cgc, sbg}\n",
    "prof = 'sbpla'\n",
    "\n",
    "config_file = sbg.Config(profile=prof)\n",
    "api = sbg.Api(config=config_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Copy a Public App\n",
    "First, we'll list all our projects. Then, we'll list the apps within the Public Reference project<sup>1</sup>. Lastly, we'll copy an app between the Public Reference project to the my\\_project. Here we also explicitly set _'limit':100_ inside the _query_. This helps speed up the auto-pagination feature within the object constructor.\n",
    "\n",
    "The critical information for this `POST` request is the **app_id**. Note, you are **not** allowed to copy the same app **and** assign the same name more than once. If you change the name, it is ok. Here we are using the App's **name**, alternatively you can match bases on the App's **ID** as shown here [apps_copyFromMyProject](apps_copyFromMyProject.ipynb). \n",
    "\n",
    "To make these results very obvious, use an empty project as your my\\_project (e.g. your <a href=projects_makeNew.ipynb> cookbook example project</a>) or add the _name_ arguement to something ridiculous like 'Orange Mocha Frapachino Maker'<sup>2</sup>\n",
    "\n",
    "### Warning on the NAME argument\n",
    "When copying apps, it is also possible to specify the **name** using\n",
    "\n",
    "```python\n",
    "    my_new_app = public_app.copy(project = my_project, name = my_app_source.name)\n",
    "```\n",
    "\n",
    "However, we caution that this can lead to conflicts with similarly named apps. Unless you are _certain_ that you want to set a specific, **different** app name, _please_ omit the name argument. It will still inherit the name from the app.\n",
    "\n",
    "```python\n",
    "    my_new_app = public_app.copy(project = my_project)\n",
    "```\n",
    "\n",
    "\n",
    "<sup>1</sup> Unlike files, apps are accesssible **both** *within* a project (here the Public Reference project) **and** by a *visibility* property (which may be set to 'public')\n",
    "\n",
    "<sup>2</sup> Note that setting the **name** of an app, actually changes the **id**. We are working on fixing this inconsistency."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment Metrics QC\n",
      "BEDTools Coverage\n",
      "BEDTools Genomecov\n",
      "BEDTools Intersect\n",
      "BEDTools Merge\n",
      "BEDTools Sort\n",
      "BWA INDEX\n",
      "BWA INDEX\n",
      "BWA MEM Bundle\n",
      "BWA MEM Bundle\n",
      "BamTools Index\n",
      "BamTools Merge\n",
      "BamTools Split\n",
      "Bcftools Annotate\n",
      "Bcftools Call\n",
      "Bcftools Convert\n",
      "Bcftools Filter\n",
      "BedGraphToBigWig\n",
      "BedToBigBed\n",
      "BigWigCorrelate\n",
      "BigWigToBedGraph\n",
      "Bowtie2 Aligner\n",
      "Bowtie2 Indexer\n",
      "BreakDancer\n",
      "BreakSeq Lite\n",
      "CNVnator\n",
      "CNVnator Analysis\n",
      "CNVnator2VCF-0-3-2\n",
      "Chimera\n",
      "ChimeraScan Gtf2Genepred\n",
      "ChimeraScan Index\n",
      "ChimeraScan Make Html\n",
      "ChimeraScan Run\n",
      "Circos\n",
      "Control-FREEC\n",
      "Control-FREEC\n",
      "Cuffdiff\n",
      "Cufflinks\n",
      "Cuffnorm\n",
      "Cuffquant\n",
      "CummeRbundQC\n",
      "DESeq2\n",
      "DeNovoGear\n",
      "Delly2\n",
      "Delly2 - Add Reference Allele\n",
      "Delly2 Workflow\n",
      "Eautils Fastq-mcf\n",
      "Exome Coverage QC 1.0\n",
      "FastQC\n",
      "FastQC Analysis\n",
      "Flexbar\n",
      "Freebayes\n",
      "Fusion Transcript Detection - ChimeraScan\n",
      "GATK ApplyRecalibration\n",
      "GATK BaseRecalibrator\n",
      "GATK CombineVariants\n",
      "GATK DiagnoseTargets\n",
      "GATK IndelRealigner\n",
      "GATK IndelRealigner Scatter\n",
      "GATK LeftAlignVariants\n",
      "GATK PrintReads\n",
      "GATK PrintReads Scatter\n",
      "GATK RealignerTargetCreator\n",
      "GATK RealignerTargetCreator Scatter\n",
      "GATK SelectVariants\n",
      "GATK UnifiedGenotyper\n",
      "GATK UnifiedGenotyper Scatter\n",
      "GATK VariantAnnotator\n",
      "GATK VariantEval\n",
      "GATK VariantFiltration \n",
      "GATK VariantRecalibrator\n",
      "GATK VariantValidationAssessor\n",
      "Graph-based whole genome sequencing analysis\n",
      "HISAT2\n",
      "HISAT2\n",
      "HISAT2 Build\n",
      "HISAT2 Build\n",
      "HISAT2 ExtractExons\n",
      "HISAT2 ExtractSNPs\n",
      "HISAT2 ExtractSpliceSites\n",
      "HISAT2 Inspect\n",
      "HTSeq-count\n",
      "Oncofuse\n",
      "Picard AddOrReplaceReadGroups\n",
      "Picard BuildBamIndex\n",
      "Picard CleanSam\n",
      "Picard CollectAlignmentSummaryMetrics\n",
      "Picard CollectGcBiasMetrics\n",
      "Picard CollectInsertSizeMetrics\n",
      "Picard CollectRnaSeqMetrics\n",
      "Picard CollectWgsMetrics\n",
      "Picard CreateSequenceDictionary\n",
      "Picard EstimateLibraryComplexity\n",
      "Picard FilterSamReads\n",
      "Picard FixMateInformation\n",
      "Picard GatherBamFiles\n",
      "Picard MarkDuplicates\n",
      "Picard MeanQualityByCycle\n",
      "Picard MergeSamFiles\n",
      "Picard QualityScoreDistribution\n",
      "Picard ReorderSam\n",
      "Picard SamToFastq\n",
      "Picard SortSam\n",
      "RNA-SeQC\n",
      "RNA-Seq Alignment - HISAT2\n",
      "RNA-Seq De Novo Assembly - Trinity 2.0.6\n",
      "RNA-Seq De Novo Assembly and Analysis - Trinity 2.0.6\n",
      "RNA-seq Alignment - STAR\n",
      "RNA-seq Alignment - STAR for TCGA PE tar\n",
      "RNA-seq Alignment - TopHat\n",
      "RNA-seq Differential Expression\n",
      "RSEM 1.2.31 Demo Workflow With Star Aligner\n",
      "RSEM BAM2WIG\n",
      "RSEM BAM2WIG\n",
      "RSEM BAM2WIG\n",
      "RSEM Calculate Expression\n",
      "RSEM Calculate Expression\n",
      "RSEM Calculate Expression\n",
      "RSEM Convert SAM\n",
      "RSEM Convert SAM\n",
      "RSEM Convert SAM\n",
      "RSEM Plot Model\n",
      "RSEM Plot Model\n",
      "RSEM Plot Model\n",
      "RSEM Plot Transcript Wiggles\n",
      "RSEM Plot Transcript Wiggles\n",
      "RSEM Plot Transcript Wiggles\n",
      "RSEM Prepare Reference\n",
      "RSEM Prepare Reference\n",
      "RSEM Prepare Reference\n",
      "RSeQC - Junction Saturation\n",
      "SAMtools BAM/CRAM <-> SAM convertor\n",
      "SAMtools BAM<->SAM Converter\n",
      "SAMtools BAM<->SAM Converter\n",
      "SAMtools Extract BAM Header\n",
      "SAMtools Extract BAM Header\n",
      "SAMtools Index BAM\n",
      "SAMtools Index BAM\n",
      "SAMtools Index BAM/CRAM\n",
      "SAMtools Index FASTA\n",
      "SAMtools Index FASTA\n",
      "SAMtools Index FASTA\n",
      "SAMtools Mpileup\n",
      "SAMtools Mpileup\n",
      "SAMtools Mpileup\n",
      "SAMtools Mpileup parallel\n",
      "SAMtools Reheader\n",
      "SAMtools Reheader\n",
      "SAMtools Reheader\n",
      "SAMtools View\n",
      "SAMtools View\n",
      "SAMtools View\n",
      "SAMtools extract SAM/BAM header\n",
      "SBG AmpliconQC: Amplicon Coverage\n",
      "SBG AmpliconQC: Coverage Heatmap\n",
      "SBG AmpliconQC: Mark Failed Amplicons\n",
      "SBG AmpliconQC: Merge Coverage BEDs\n",
      "SBG Bedpe4Oncofuse\n",
      "SBG ChimeraScan4Circos\n",
      "SBG Compressor\n",
      "SBG Coverage\n",
      "SBG Decompressor\n",
      "SBG FASTA Indices\n",
      "SBG FASTQ Merge\n",
      "SBG FASTQ Quality Adjuster\n",
      "SBG FASTQ Quality Converter\n",
      "SBG FASTQ Quality Detector\n",
      "SBG FASTQ Split\n",
      "SBG FastQC Beautifier\n",
      "SBG FastQC Extract\n",
      "SBG FilterFasta\n",
      "SBG Flatten\n",
      "SBG FlattenLists\n",
      "SBG GC Content\n",
      "SBG GC Coverage\n",
      "SBG Genome Coverage\n",
      "SBG Group Input\n",
      "SBG Html2b64\n",
      "SBG Oncofuse4Circos\n",
      "SBG PED File Creator\n",
      "SBG Pair Contigs\n",
      "SBG Pair FASTQs by Metadata\n",
      "SBG Pass Contigs\n",
      "SBG Pass Intervals\n",
      "SBG Picard GatherBamFiles Sorted\n",
      "SBG Prepare GATK VariantRecalibrator Resource\n",
      "SBG Prepare Intervals\n",
      "SBG Push2Solvebio\n",
      "SBG Report Renderer 2\n",
      "SBG SAMtools Merge Mpileup\n",
      "SBG Split BED\n",
      "SBG SplitFasta\n",
      "SBG Text2Html\n",
      "SBG UCSC-B37 BED Converter\n",
      "SBG Unpack FASTQs\n",
      "SBG Untar fasta\n",
      "SBG VCF Reorder\n",
      "SBG Vcf2Json\n",
      "SRA Toolkit fastq-dump\n",
      "SRA Toolkit sam-dump\n",
      "STAR\n",
      "STAR\n",
      "STAR Genome Generate\n",
      "STAR Genome Generate\n",
      "STAR-Fusion\n",
      "STARlong\n",
      "Sambamba Index\n",
      "Sambamba Merge\n",
      "Sambamba Sort\n",
      "Sambamba View\n",
      "Samtools dict\n",
      "Samtools sort\n",
      "SnpEff\n",
      "SnpEff Download\n",
      "SnpEff Download\n",
      "SnpEff adjusted\n",
      "SnpSift Annotate \n",
      "SnpSift Annotate \n",
      "SnpSift dbNSFP\n",
      "SnpSift dbNSFP\n",
      "SolveBio Integration\n",
      "SomaticSniper\n",
      "SomaticSniper + Filtering\n",
      "SomaticSniper BAM Readcount\n",
      "SomaticSniper FP Filter\n",
      "SomaticSniper HC Filter\n",
      "SomaticSniper Prepare for BAM Readcount\n",
      "SomaticSniper SNP Filter\n",
      "Tabix BGZIP\n",
      "Tabix Index\n",
      "Targeted Capture Analysis - BWA + GATK 2.3.9-Lite \n",
      "TopHat2\n",
      "Torrent Mapping Alignment Program Index\n",
      "Torrent Mapping Alignment Program Map1\n",
      "Torrent Mapping Alignment Program Map2\n",
      "Torrent Mapping Alignment Program Map3\n",
      "Torrent Mapping Alignment Program Map4\n",
      "Torrent Mapping Alignment Program MapVSW\n",
      "Torrent Variant Caller\n",
      "Trinity Abundance Estimates to Matrix 2.0.6\n",
      "Trinity Align and Estimate Abundance 2.0.6\n",
      "Trinity Analyze DE 2.0.6\n",
      "Trinity Assembler 2.0.6\n",
      "Trinity Component Distribution 2.0.6\n",
      "Trinity Count Features 2.0.6\n",
      "Trinity Define Clusters 2.0.6\n",
      "Trinity Run DE Analysis 2.0.6\n",
      "Trinity Stats 2.0.6\n",
      "VCFtools Concat\n",
      "VCFtools Hardy\n",
      "VCFtools Hwe\n",
      "VCFtools Isec\n",
      "VCFtools Merge\n",
      "VCFtools Sort\n",
      "VCFtools Subset\n",
      "VCFtools Venn2\n",
      "VCFtools Venn3\n",
      "VarScan2 CopyCaller\n",
      "VarScan2 CopyNumber\n",
      "VarScan2 Filter\n",
      "VarScan2 ProcessSomatic\n",
      "VarScan2 Somatic\n",
      "VarScan2 Somatic parallel\n",
      "VarScan2 SomaticFilter\n",
      "VarScan2 SomaticFilter parallel \n",
      "VarScan2 Workflow from BAM\n",
      "Variant Effect Predictor\n",
      "Whole Exome Analysis - BWA + GATK 2.3.9-Lite (with Metrics)\n",
      "Whole Genome Analysis - BWA + GATK 2.3.9-Lite (with Metrics)\n",
      "eXpress\n",
      "eXtasy\n"
     ]
    }
   ],
   "source": [
    "# (optional) Print names of all Public Apps\n",
    "A = [a.name for a in api.apps.query(visibility='public', limit=100).all()]\n",
    "A.sort()\n",
    "\n",
    "for a in A:\n",
    "    print a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "App (Whole Exome Analysis - BWA + GATK 2.3.9-Lite (with Metrics)) does not exist in Project (Hopping like the Kangaroos); copying now\n",
      "Sucessfully copied one app!\n"
     ]
    }
   ],
   "source": [
    "# [USER INPUT] Set project name; project (p_) and file (f_) indices here:\n",
    "my_project_name = 'Keep on Smiling'\n",
    "a_name = 'Whole Exome Analysis - BWA + GATK 2.3.9-Lite (with Metrics)'\n",
    "\n",
    "\n",
    "# pull out target project\n",
    "my_project = [p for p in api.projects.query(limit=100).all() \\\n",
    "              if p.name == my_project_name]\n",
    "\n",
    "# Double-check that target project exists\n",
    "if not my_project:\n",
    "    print('Target project (%s) not found, check spelling' % my_project_name)\n",
    "    raise KeyboardInterrupt\n",
    "else:\n",
    "    my_project = my_project[0]\n",
    "    \n",
    "my_apps = api.apps.query(project = my_project.id, limit=100)\n",
    "public_app = [a for a in api.apps.query(visibility='public', limit=100).all() \\\n",
    "                 if a.name == a_name][0]\n",
    "\n",
    "duplicate_app = [a for a in my_apps.all() if a.name == public_app.name]\n",
    "\n",
    "if duplicate_app:\n",
    "    print('App already exists in second project, please try another app')\n",
    "else:\n",
    "    print('App (%s) does not exist in Project (%s); copying now' % \\\n",
    "          (a_name, my_project.name))\n",
    "    \n",
    "    new_app = public_app.copy(project = my_project.id)\n",
    "        \n",
    "    # re-list apps in target project to verify the copy worked\n",
    "    my_app_names = [a.name for a in \\\n",
    "                    api.apps.query(project = my_project.id, limit=100).all()]\n",
    "    \n",
    "    if a_name in my_app_names:\n",
    "        print('Sucessfully copied one app!')\n",
    "    else:\n",
    "        print('Something went wrong...')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Additional Information\n",
    "Detailed documentation of this particular REST architectural style request is available [here](http://docs.sevenbridges.com/docs/copy-an-app)."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
